### PR TITLE
[NEW]: TD3D segmentation head

### PR DESCRIPTION
**Pull request by col14m**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Review by filaPro — CHANGES_REQUESTED** [review]
May be one more sentences in comment about our changes.

File: projects/TD3D/td3d/td3d.py [hunk 1]
```diff
@@ -129,6 +129,14 @@ def loss(self, batch_inputs_dict: dict, batch_data_samples: SampleList,
         annotated_points = []
         for i in range(len(batch_data_samples)):
             gt_pts_seg = batch_data_samples[i].gt_pts_seg
+
+            gt_pts_seg.pts_instance_mask[gt_pts_seg.pts_semantic_mask == 18] = -1 #todo: why 18?
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d.py]: move this `for` to the head. we even don't know num_classes here. and remove todo

File: projects/TD3D/configs/td3d.py [hunk 2]
```diff
@@ -25,8 +25,17 @@
         pts_center_threshold=8),
     seg_head=dict(
         type='TD3DSegmentationHead',
-        voxel_size=0.02),
-    train_cfg=dict(num_proposals=2),
+        voxel_size=0.02,
+        target_iou_thr=0.25,
+        padding=0.08,
+        min_pts_threshold=10,
+        n_classes=18,
```
Comment (filaPro) [inline projects/TD3D/configs/td3d.py]: num_classes

File: projects/TD3D/td3d/td3d_segmentation_head.py [hunk 3]
```diff
@@ -14,47 +14,466 @@
 
 from mmengine.model import BaseModule
 from mmengine.structures import InstanceData
-from mmdet3d.registry import MODELS
+from mmdet3d.registry import MODELS, TASK_UTILS
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d_segmentation_head.py]: do we need TASK_UTILS?

File: projects/TD3D/td3d/td3d_segmentation_head.py [hunk 4]
```diff
@@ -14,47 +14,466 @@
...
-
+from mmdet3d.structures import BaseInstance3DBoxes
+from mmdet3d.structures import AxisAlignedBboxOverlaps3D
 
 @MODELS.register_module()
 class TD3DSegmentationHead(BaseModule):
+    r"""Refinement head (second stage) of 
+            `TD3D <https://arxiv.org/abs/2302.02871>`_.
+    Args:
+        n_classes (int): Number of classes.
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d_segmentation_head.py]: num_classes

File: projects/TD3D/td3d/td3d_segmentation_head.py [hunk 5]
```diff
@@ -14,47 +14,466 @@
...
+        voxel_size (float): Voxel size in meters.
+        target_iou_thr (float): IoU threshold for
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d_segmentation_head.py]: can we move it to train_cfg?

File: projects/TD3D/td3d/td3d_segmentation_head.py [hunk 6]
```diff
@@ -14,47 +14,466 @@
...
+        seg_targets[seg_targets == -1] = self.n_classes
+        seg_loss = self.multiclass_loss(seg_preds, seg_targets.long())
+
+        inst_loss = self.binary_loss(inst_preds, labels)
+        return inst_loss, seg_loss
+  
     def predict(self,
                 x: SparseTensor, 
-                field: TensorField,
+                points: TensorField,
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d_segmentation_head.py]: why points? we use this name several times in this head and in detection head. i think we can call TensorField as field here

File: projects/TD3D/td3d/td3d_segmentation_head.py [hunk 7]
```diff
@@ -14,47 +14,466 @@
...
+        dx_min = points[..., 0] - boxes[..., 0] + boxes[..., 3] / 2
+        dx_max = boxes[..., 0] + boxes[..., 3] / 2 - points[..., 0]
+        dy_min = points[..., 1] - boxes[..., 1] + boxes[..., 4] / 2
+        dy_max = boxes[..., 1] + boxes[..., 4] / 2 - points[..., 1]
+        dz_min = points[..., 2] - boxes[..., 2] + boxes[..., 5] / 2
+        dz_max = boxes[..., 2] + boxes[..., 5] / 2 - points[..., 2]
+        return torch.stack((dx_min, dx_max, dy_min, dy_max, dz_min, dz_max), dim=-1)
+
+    def _loss(self, cls_preds: Tensor, targets: Tensor, 
+              v2r: Tensor, r2scene: Tensor,
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d_segmentation_head.py]: voxel2roi, roi2scene and in other places

File: projects/TD3D/td3d/td3d_segmentation_head.py [hunk 8]
```diff
@@ -14,47 +14,466 @@
...
-                field: TensorField,
+                points: TensorField,
                 proposals: List[InstanceData],
                 batch_data_samples: SampleList,
                 **kwargs) -> List[PointData]:
+        """Generate masks for all scenes.
+
+        Args:
+            x (SparseTensor): Input feature map.
+            points (TensorField): src points for inverse mapping
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d_segmentation_head.py]: TensorField containing source points for ...

File: projects/TD3D/td3d/td3d_segmentation_head.py [hunk 9]
```diff
@@ -14,47 +14,466 @@
...
+
+        Args:
+            x (SparseTensor): Input feature map.
+            points (TensorField): src points for inverse mapping 
+            proposals (List[InstanceData]): Predicted bounding boxes for all
+                scenes in a mini-batch.
+            batch_data_samples (SampleList): Batch of meta info.
+
+        Returns:
+            results (List[PointData]): final predicted masks.
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d_segmentation_head.py]: Predicted instances with masks, scores and labels.

File: projects/TD3D/td3d/td3d_segmentation_head.py [hunk 10]
```diff
@@ -14,47 +14,466 @@
...
+                idxs=idxs[v2scene == i],
+                v2r=v2r[v2scene == i],
+                scores=scores[i],
+                labels=labels[i],
+                inverse_mapping=inv_mapping)
+            results.append(result)
+
         return results
+
+    # per scene
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d_segmentation_head.py]: do we need this comment?

File: projects/TD3D/td3d/td3d_segmentation_head.py [hunk 11]
```diff
@@ -14,47 +14,466 @@
...
+                              inverse_mapping: Tensor) -> PointData:
+        """Get instances masks for one scene.
+        
+        Args:
+            cls_preds (Tensor): Masks predictions for each proposal.
+            idxs (Tensor): Voxels idxs in source voxel space.
+            v2r (Tensor): Voxel to roi mapping.
+            scores (Tensor): Scores for proposals.
+            labels (Tensor): Labels for proposals.
+            inverse mapping (Tensor): voxel to point mapping.
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d_segmentation_head.py]: fix `_` and start with capital letter

File: projects/TD3D/td3d/td3d_segmentation_head.py [hunk 12]
```diff
@@ -14,47 +14,466 @@
...
+        
+        Returns:
+            (PointData): final instances.
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d_segmentation_head.py]: Why ()? and start with capital letter

File: projects/TD3D/td3d/td3d_segmentation_head.py [hunk 13]
```diff
@@ -14,47 +14,466 @@
...
+        idxs_expand = idxs.unsqueeze(-1).expand(idxs.shape[0], 
+                                                n_rois).long()
+        bin_preds_exp = bin_cls_preds.unsqueeze(-1).expand(bin_cls_preds.shape[0], 
+                                                                        n_rois)
+        cls_preds[cls_preds <= self.test_cfg.seg_score_thr] = 0
+        cls_preds_expand = cls_preds.unsqueeze(-1).expand(cls_preds.shape[0],
+                                                         n_rois)
+        idxs_expand[~v2r_one_hot] = inverse_mapping.max() + 1
+
+        voxels_masks = idxs.new_zeros(inverse_mapping.max() + 2,
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d_segmentation_head.py]: may be a comment why + 2 here?

File: projects/TD3D/td3d/td3d_segmentation_head.py [hunk 14]
```diff
@@ -14,47 +14,466 @@
...
+        assert torch.all(torch.unique(v2r) == torch.arange(0, 
+                                            v2r.max() + 1).to(v2r.device))
+
+        cls_preds = cls_preds.sigmoid()
+        bin_cls_preds = cls_preds > self.test_cfg.seg_score_thr
+        v2r_one_hot = torch.nn.functional.one_hot(v2r).bool()
+        n_rois = v2r_one_hot.shape[1]
+        idxs_expand = idxs.unsqueeze(-1).expand(idxs.shape[0], 
+                                                n_rois).long()
+        bin_preds_exp = bin_cls_preds.unsqueeze(-1).expand(bin_cls_preds.shape[0],
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d_segmentation_head.py]: why _exp and _exapand in these lines? may be even indexes and bin_preds

File: projects/TD3D/td3d/td3d_segmentation_head.py [hunk 15]
```diff
@@ -14,47 +14,466 @@
...
+        Args:
+            x (SparseTensor): Input feature map for proposals generation.
+            proposals (List[InstanceData]): Predicted bounding boxes for all
+                scenes in a mini-batch. Proposals[i].labels_3d contains assigned 
+                idxs for gt bboxes from the first stage.
+            pts_targets (Tensor): voxel-wise instance and semantic markup. 
+            batch_data_samples (SampleList): Batch of meta info.
+
+        Returns:
+            dict: instance and semantic losses.
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d_segmentation_head.py]: capital lettter. and in other places

File: projects/TD3D/td3d/td3d_segmentation_head.py [hunk 16]
```diff
@@ -14,47 +14,466 @@
...
+                 proposals: List[BaseInstance3DBoxes]) -> Tuple:
+        """Extract proposals and pass them through unet.
+        
+        Args:
+            x (SparseTensor): Input feature map for proposals generation.
+            pts_targets (Tensor): Voxel-wise instance and semantic markup.
+            proposals (List[BaseInstance3DBoxes]): Predicted bounding boxes for all
+                scenes in a mini-batch.
+        
+        Returns:
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d_segmentation_head.py:170]: Incorrect docstring format. Please look at https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/mmdet3d/models/detectors/imvoxelnet.py#L81 .

File: projects/TD3D/td3d/td3d_segmentation_head.py [hunk 17]
```diff
@@ -14,47 +14,466 @@
...
+            preds (Tensor): Extracted feature maps for all valid proposals.
+            targets (Tensor): Extracted voxel-wise instance and semantic markup 
+                for all valid proposals.
+            v2r (Tensor): voxel to roi mapping.
```
Comment (filaPro) [inline projects/TD3D/td3d/td3d_segmentation_head.py]: voxel2roi and in other places

**Review by filaPro — CHANGES_REQUESTED** [review]

File: projects/TD3D/td3d/instance_seg_eval.py [hunk 18]
```diff
@@ -27,8 +27,7 @@ def aggregate_predictions(masks, labels, scores, valid_class_ids):
         label = label.clone().numpy()
         score = score.clone().numpy()
         info = dict()
-        n_instances = mask.max() + 1
-        for i in range(n_instances):
+        for i in range(mask.shape[0]):
```
Comment (filaPro) [inline projects/TD3D/td3d/instance_seg_eval.py:31]: Please write one more sentences about our changes in the comment before this function

File: projects/TD3D/td3d/td3d_segmentation_head.py (hunk 14 above)
Comment (col14m) [reply projects/TD3D/td3d/td3d_segmentation_head.py]: I'm not sure I got you right, so you will need to look at my changes.

File: projects/TD3D/td3d/td3d.py (hunk 1 above)
Comment (col14m) [reply projects/TD3D/td3d/td3d.py]: I am not sure about it. It's very inconveniently to do it in the head, because markup is already combined into a batch there. Maybe we can do it during data processing stage? For example, we can create new augmentation
